library/common/http/dispatcher.cc

```diff
@@ -49,75 +65,152 @@ void Dispatcher::DirectStreamCallbacks::onHeaders(HeaderMapPtr&& headers, bool e
+    ENVOY_LOG(debug, "[S{}] complete stream", direct_stream_.stream_handle_);
+    if (direct_stream_.dispatchable(true)) {
+      ENVOY_LOG(debug, "[S{}] dispatching to platform complete stream",
+                direct_stream_.stream_handle_);
+      bridge_callbacks_.on_complete(bridge_callbacks_.context);
+    }
+    direct_stream_.closeRemote(true);
+  }
 }
 
+Stream& Dispatcher::DirectStreamCallbacks::getStream() { return direct_stream_; }
+
 void Dispatcher::DirectStreamCallbacks::onReset() {
-  ENVOY_LOG(debug, "[S{}] remote reset stream", stream_handle_);
+  ENVOY_LOG(debug, "[S{}] remote reset stream", direct_stream_.stream_handle_);
   envoy_error_code_t code = error_code_.value_or(ENVOY_STREAM_RESET);
   envoy_data message = error_message_.value_or(envoy_nodata);
-  bridge_callbacks_.on_error({code, message}, bridge_callbacks_.context);
-  // Very important: onComplete and onReset both clean up stream state in the http dispatcher
-  // because the underlying async client implementation **guarantees** that only onComplete **or**
-  // onReset will be fired for a stream. This means it is safe to clean up the stream when either of
-  // the terminal callbacks fire without keeping additional state in this layer.
-  http_dispatcher_.cleanup(stream_handle_);
+  // Note: in the case that we received a complete remote response but envoy is resetting the stream
+  // due to an incomplete local request this on_error call will happen after an on_complete.
+  // FIXME: before merge. we still have to think these scenarios a bit more carefully.
+  // FIXME: before merge. After some discussion of this we decided to
+  // move atomic platform state, contracts, and cancellation down here.
+  // Will do that ASAP after getting the HCM out of envoy.
+  if (direct_stream_.dispatchable(true)) {
+    ENVOY_LOG(debug, "[S{}] dispatching to platform remote reset stream",
+              direct_stream_.stream_handle_);
+    bridge_callbacks_.on_error({code, message}, bridge_callbacks_.context);
+  }
+  // FIXME: before merge. this can't be right because then we could be potentially calling cleanup
+  // twice on a stream. For instance on an inline reset? Not for the case with the local_closed_
+  // false but remote_closed_ true. Because in that case the cleanup call is guarded by complete(),
+  // right?
+  http_dispatcher_.cleanup(direct_stream_.stream_handle_);
+}
+
+void Dispatcher::DirectStreamCallbacks::onCancel() {
+  // This call is guarded at the call-site @see Dispatcher::DirectStream::resetStream().
+  // Therefore, it is dispatched here without protection.
+  ENVOY_LOG(debug, "[S{}] dispatching to platform cancel stream", direct_stream_.stream_handle_);
+  bridge_callbacks_.on_cancel(bridge_callbacks_.context);
+
+  // FIXME: before merge. what about clean up? It all depends on if the reset we send down causes a
+  // reset coming up, which will not fire a callback, but will result in stream clean up.
```

Comment:
I'd defer changing logic here, since we know we plan to move it down into the core layer.